.github/workflows/pr_require_single_commit.yml

```diff
@@ -0,0 +1,20 @@
+name: Require PR has a single commit
+permissions: read-all
+on:
+  pull_request:
+    branches:
+      - main
+jobs:
+  require-single-commit:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out repository code
+        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # pin@v3
+
+      - name: Assert single commit
+        run: |
+          commits=${{ github.event.pull_request.commits }}
+          if (( ${commits} > 1 )); then
+            echo "Expected PR to have a single commit, but found ${commits}; google/heir requires a single commit per PR because Google's internal/external synchronization tooling does not support squashing, and instead would rebase all commits into the main branch."
```

Comment:
Try this for workflow errors! https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-creating-an-annotation-for-an-error

Reply:
oh the `core.setFailed` is from their typescript toolkit, in the workflow bash you can just do `::error::expected PR...`

Reply:
Just went with `error`.